### ISSUE TITLE
When using git.present, salt $host state.highstate output declares a change every time, even if the repository is present already.

### ISSUE THREAD

**Issue by the-l-a-cat** [saltstack/salt#15517]
When using git, I notice state.highstate output always says:

```
          ID: zsh
    Function: git.present
        Name: https://github.com/zsh-users/zsh-syntax-highlighting
      Result: True
     Comment: Initialized repository https://github.com/zsh-users/zsh-syntax-highlighting
     Changes:   
              ----------
              new repository:
                  https://github.com/zsh-users/zsh-syntax-highlighting
```

This is the relevant sls code:
```
 zsh:
    pkg.installed:
        - require:
            - git: zsh 
    git.present:
        - name: https://github.com/zsh-users/zsh-syntax-highlighting
        - target: /usr/share/zsh/plugins/zsh-syntax-highlighting
```

Of course I mean that repository is already cloned into the correct place with the first iteration of state.highstate call. It does not appear to me that any files are actually being overwritten, as one may suggest looking at the output. I believe the correct behaviour to be to make no changes and indicate that no changes were made, and that currently that indication lies.

```
 # salt-call --versions-report
           Salt: 2014.1.10
         Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
 msgpack-python: 0.3.0
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
         PyYAML: 3.10
          PyZMQ: 14.0.1
            ZMQ: 4.0.4
 # uname -a
 Linux host 3.13.0-24-generic #46-Ubuntu SMP Thu Apr 10 19:11:08 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by eliasp** [comment]
`2014.7.0` is affected in the same way as `2014.1.10`.

**Comment by basepi** [comment]
@eliasp Thanks for the update. I don't think anyone's had time to investigate this issue, so I think that's expected.

**Comment by aneeshusa** [comment]
Started taking a look into this. As far as I can tell from the [docs](http://docs.saltstack.com/en/latest/ref/states/all/salt.states.git.html#salt.states.git.present) and the [code](https://github.com/saltstack/salt/blob/develop/salt/states/git.py#L410), git.present does not actually support target as an option. Name is the destination of the directory with the git repo, not where to clone the git repo from.

@the-l-a-cat and @eliasp,  I'm a little confused as to exactly how you want the target option to work. If you want the repo to be cloned from the external target if it doesn't already exist, I believe [git.latest](http://docs.saltstack.com/en/latest/ref/states/all/salt.states.git.html#salt.states.git.latest) will do what you want. However, git.latest will always update the repo, but you state "the correct behaviour [is] to make no changes". Could you clarify as to what you're expecting? 

However, there still is a bug. git.present will check if "https://github.com/zsh-users/zsh-syntax-highlighting" is a directory, which is isn't, so it will always go ahead and call git.init. However, git.init delegates out to cmd.run_all, which seeing no current working directory, sets it to the home directory of the user running salt, and calls git init in that directory. git init will then make the directory (assuming running as root): "/root/https:/github.com/zsh-users/zsh-syntax-highlighting". This doesn't match the directory that git.present is checking, which is why git.present always seems to reinitialize the repository.

The bug is caused when the name is a relative path instead of an absolute one, so I think the best way to fix this is using os.path.isabs in git.present before checking if the directory exists, prepending the correct home directory before relative paths and leaving absolute paths alone. I can go ahead and put together a PR if that sounds appropriate.

**Comment by jfindlay** [comment]
@terminalmage, should this also be considered in the git exec/state module rewrite?

**Comment by terminalmage** [comment]
Yeah, this should no longer be an issue after the rewrite, but I will confirm here and if not make sure that it is fixed. Thanks for the heads-up.

**Comment by terminalmage** [comment]
Confirmed fixed in #26431, fix will be in 2015.8.0.

**Comment by terminalmage** [comment]
Fixed issue number in previous comment.
